### ISSUE TITLE
#hospiceOverflow: interrupt during process termination creates a zombie

### ISSUE THREAD

**Issue by jgfoster** [dolphinsmalltalk/Dolphin#72]
Sometimes when I close an application window I get a gpFault. The following screen show shows the debugger and stack. It appears that the uiIdle method is iterating over the topLevelWindows when one of the windows goes away. Particularly surprising is that the receiver of #primTerminate is SmallInteger and it is implementing a method in Process. This looks to me like a stack corruption.

![screen shot 2016-02-11 at 6 15 08 pm](https://cloud.githubusercontent.com/assets/1577872/12997171/84d2f49c-d0eb-11e5-8249-ff970a603e38.png)

**Comment by blairmcg** [comment]
@jgfoster, when you say "closing an app window" is this running in a development system, or not? And is the window (or the session manager if in a deployed app) explicitly performing at quit:?
In an application session manager, when the idle loop detects that there are no top-level windows 
What you might be seeing here is the problem with the quitPrimitive attempting to throw an exception across a user/kernel mode transition, or it might just be a bug in the terminate primitive. There is a bug in the VM's handling of uncaught exceptions encountered when a primitive is running. When it does so it fails the primitive, causing the Smalltalk backup code in the primitive method to run, and it also sends an interrupt that you see as the vmi:list:no:with: frame, and it is the handler for that "GPF" interrupt that shows the walkback. The VM bug is that sometimes the stack pointer saved down for the primitive frame is incorrect (I think it probably depends on the particular primitive where the fault occurred), so the debugger sees the wrong receiver (in this case probably one of the arguments). In other words the stack state you see is a result of a bug in the handling of the exception/gpf, and probably not the cause of it.
I haven't seen this particular issue in the dev system, although I do occasionally encounter a problem with the uiIdle loop attempting to update the command status in a window after it has been partially closed, which causes a walkback attempting to perform some operation on a dead window.

**Comment by jgfoster** [comment]
@blairmcg, thanks for following up on this. Yes, this is in an development system (hence I was able to get a walkback and a debugger). It might also be in a runtime but I'm not sure. Further investigation shows that it coincides with the existence of dead processes (as shown in the process monitor).

I've got a reproducible case but I haven't been able to narrow it down to a test case that can be loaded into DPRO. It seems to involve terminating a process that is waiting on a Delay. If I don't terminate the process (but simply let it complete), then there is no dead process. 

When running a VMDebug configuration I sometimes stumble across one of [these](https://github.com/dolphinsmalltalk/DolphinVM/issues/12) issues and I wonder if it is related.

I've spent several hours trying to narrow in on the problem. Right now I'm analyzing the Process switch messages generated by the tracing code in the VM. One interesting item is "WARNING: Switching processes with interrupts disabled."

If you have any other suggestions I'd welcome them.

**Comment by jgfoster** [comment]
After many more hours of tracing I've narrowed it down a bit. In the following we start and end with the **Idler** process. **Main** creates **proc-A** which creates **proc-B**. **proc-B** sits on a Delay and is terminated during the Delay. During the termination process, **Main** goes on to do some activity that creates garbage and is interrupted with a **#hospiceOverflow:**. After that we get **Warning: Switching processes with interrupts disabled** and the Process Monitor shows **proc-B** as dead but hanging around (a *zombie*?). Subsequent activity (typically closing any window, including a confirmation dialog) ends up with the corrupt stack shown above.

My application work-around is to avoid terminating **proc-B** and let it run to completion normally. With this change there are no more zombie processes and no GPF walkbacks. Of course, there are times when it would be useful to terminate processes, so this isn't a perfect solution. 

As best I can tell, **proc-B** is not referenced from the image (#allReferences returns an empty collection), so it must be referenced from the VM. I suspect that the **#hospiceOverflow:** interrupt prevented the process termination from completing. When I modify my application so that there is not as much garbage generated (extract fewer elements from an XML Document), then there is no zombie process. If I extract 5 elements, things are fine. If I extract 6 elements then I get the zombie. This has, however, been difficult to extract into an isolated TestCase (without my application code).

Following is the trace output (edited for clarity):
* 3287: Switching  
 from a Process('Idler' [ACTIVE] in Semaphore>>set)  
 to a Process('Main' in Semaphore>>wait)  
* 'Dolphin7.exe' (Win32): Loaded '\\vmware-host\Shared Folders\jfoster On My Mac\Documents\Dolphin\DolphinCR7.DLL'. Symbols loaded.  
* BasepIsRemovableMedia: Host device is removable, Shim cache deactivated  
* Interpreter: Yield  
 from a Process('Main' [ACTIVE] in Mutex>>critical:)  
 to a Process('proc-A' in BlockClosure>>newProcess)  
* 3288: Switching  
 from a Process('Main' [ACTIVE] in Mutex>>critical:)  
 to a Process('proc-A' in BlockClosure>>newProcess)  
* 3289: Switching  
 from a Process('proc-A' [ACTIVE] in Process>>shutdown0)  
 to a Process('proc-B' in BlockClosure>>newProcess)  
* 3290: Switching  
 from a Process('proc-B' [ACTIVE] in Semaphore>>critical:atPriority:)  
 to a Process('Main' in Mutex>>critical:)  
* Interrupting  
 a Process('proc-B' in Semaphore>>critical:atPriority:)  
 with 1:nil (**#terminate**)  
* 3291: Switching  
 from a Process('Main' [ACTIVE] in GciSession>>incrementEventCount)  
 to a Process('proc-B' in Semaphore>>critical:atPriority:)  
* 3292: Switching  
 from a Process('proc-B' [ACTIVE] in ProcessorScheduler>>enableInterrupts0)  
 to a Process('Main' in GciSession>>incrementEventCount)  
* 'Dolphin7.exe' (Win32): Loaded 'C:\Windows\System32\msxml3.dll'. Cannot find or open the PDB file.  
* Interpreter: Yield  
 from a Process('Main' [ACTIVE] in Mutex>>critical:)  
 to a Process('proc-B' in ProcessorScheduler>>enableInterrupts)  
* 3293: Switching  
 from a Process('Main' [ACTIVE] in Mutex>>critical:)  
 to a Process('proc-B' in ProcessorScheduler>>enableInterrupts)  
* Interrupting  
 a Process('Main' in Mutex>>critical:)  
 with 18:64 (**#hospiceOverflow:**)  
* 3294: Switching  
 from a Process('proc-B' [ACTIVE] in ProcessorScheduler>>terminateActive)   
 to a Process('Main' in Mutex>>critical:)  
* **WARNING: Switching processes with interrupts disabled**  
* 3295: Switching  
 from a Process('Main' [ACTIVE] in Mutex>>critical:)  
 to a Process('proc-B' in ProcessorScheduler>>terminateActive)  
* 3296: Switching  
 from a Process('proc-B' [ACTIVE] in Process>>shutdown)  
 to a Process('Main' in Mutex>>critical:)  
* 3297: Switching  
 from a Process('Main' [ACTIVE] in Semaphore>>wait)  
 to a Process('Finalizer' in Semaphore>>wait) 
* 3298: Switching  
 from a Process('Finalizer' [ACTIVE] in Semaphore>>wait)  
 to a Process('Idler' in Semaphore>>set)

**Comment by jgfoster** [comment]
The following will generally leave a zombie process and might even cause Dolphin itself to crash:

  | proc array |
  proc := [(Delay forSeconds: 1) wait] fork.
  Processor yield.
  array := Array new: 100.
  1 to: 1000 do: [:i | 
    | j |
    j :=  i \\ 100 + 1.
    (array at: j) ifNotNil: [:fs | fs close].
    array at: j put: (FileStream read: 'DPRO.chg').
    i == 200 ifTrue: [proc terminate].
  ].

**Comment by blairmcg** [comment]
Excellent effort diagnosing that @jgfoster. It begins to make sense to me that there may be a problem with an interrupt being sent to the process when it is running the Process>>shutdown method. My first reaction was that this should be avoided, but I think that would be just masking a problem in either the interrupt delivery code, or the terminate primitive. I need to have a think about the right solution here and try and understand the fault.
In the meantime are you able to work around the issue by avoiding explicit termination of the process? e.g. set a flag that it looks at when it wakes up.

**Comment by jgfoster** [comment]
@blairmcg, Yes, I have a perfectly fine work-around for my specific use-case. My primary interest (and why spent so much time) was in giving you enough information to understand what is almost certainly a real—but rare—race condition (and having access to the VM source allowed me to do that investigation). These problems almost always seem more obvious in retrospect; the idea that interrupting a shutdown could cause an incomplete shutdown seems pretty obvious now, but I recognize that the solution may require some care. Take the time needed.

**Comment by blairmcg** [comment]
@jgfoster, I have found a specific issue with the VM code that removes a process from the list it is waiting on when it is terminated (if it is not the active process). This code used to be correct, but at some point a declaration of a local as a reference to a pointer (i.e. *&) was changed to just a pointer. This was probably done when direct use of some instance fields was replaced with an inline function (I can't remember why that was done). But cutting a long story short, this resulted in the Process->next pointer not being nil'd when your test script is run and the process waiting on the delay was terminated. Although I wasn't able to repro the behavior you are seeing, I did find that having saved the image after running your script a few times, I could not restart it. Also this could explain the zombie processes since the linked list of processes was not being maintained correctly.
